pandas/io/common.py

```diff
@@ -611,42 +612,66 @@ def get_handle(
         encoding = "utf-8"
 
     # Convert pathlib.Path/py.path.local or string
-    path_or_buf = stringify_path(path_or_buf)
-    is_path = isinstance(path_or_buf, str)
-    f = path_or_buf
+    handle = stringify_path(path_or_buf)
 
     compression, compression_args = get_compression_method(compression)
-    if is_path:
-        compression = infer_compression(path_or_buf, compression)
+    compression = infer_compression(handle, compression)
+
+    # memory mapping needs to be the first step
+    memory_map &= hasattr(handle, "fileno") or isinstance(handle, str)
+    if memory_map:
+        if isinstance(handle, str):
```

Comment:
i think might be ok to make a separate function for this

`handle, handles = maybe_memory_map(....)`

not sure if that is more clear